### PR TITLE
[Bugfix][Quantization] Stop sleep(level=2) from zeroing compressed-tensors KV scales

### PR DESCRIPTION
**Pull request by AlanFokCo**
## Purpose

`sleep(level=2)` zeroes the calibrated KV-cache scales of compressed-tensors
KV-FP8 checkpoints, so the first request after a weights-only reload — the shape an
RL weight-sync uses — returns **NaN logprobs and a wrong first token**.

Level 2 discards the weights allocation without keeping a copy, and the only
generic safety net is the `named_buffers()` snapshot in `Worker.sleep()` /
`Worker.wake_up()` (`vllm/v1/worker/gpu_worker.py:253`, `:291`) — that file has no
`named_parameters` walk. `CompressedTensorsKVCacheMethod.process_weights_after_loading()`
rebinds `_q_scale`/`_k_scale`/`_v_scale` onto the `nn.Parameter` placeholders,
moving those names out of `_buffers`, so the snapshot never sees them and they come
back as blank pages reading `0.0`. `_prob_scale`, never rebound, survives — that
asymmetry pinpoints the cause, and every other KV-cache quant method is unaffected
because it `.copy_()`s into the existing buffer (`kv_cache.py:152/156/157/195`).

**Fix (+10/−3, one file):** `register_buffer()` the three scales — the form
`set_default_quant_scales(register_buffer=True)` created them in
(`vllm/model_executor/layers/attention/attention.py:129-135`). `register_buffer()`
rather than `.copy_()` because ATTN_HEAD loads `num_kv_heads` scales into a buffer
created 0-dim; `persistent=True` because `sharded_state_loader.py:162` needs the
keys. Shape, dtype, device, virtual address and the `state_dict` key set are
unchanged (measured: same keys, shapes and dtypes, 0 pointer moves). The single exit
covers TENSOR and ATTN_HEAD by measurement, MLA and the draft model statically (same
`_init_kv_cache_quant`, same buffers snapshotted) — we have not run MLA or
spec-decode.

We fixed the quant method rather than the snapshot because compressed-tensors is the
only one that breaks that buffer invariant; a worker-level `named_parameters()` net
would also cover the W8A8 case below, but it carries a host-memory cost and a policy
question, so it belongs in its own change.

## Duplicate-work check

Searched open PRs/issues for compressed-tensors KV scales, `sleep(level=2)`,
`_q/_k/_v_scale`, the `named_buffers` snapshot and `register_buffer`: nothing open
addresses this. Nearest hits #38975 (`_prob_scale_float` sync, no sleep involved)
and #55206 (release pinned host memory after `wake_up`). Related RFCs #48310, #48312.

## Test plan

```bash
pytest -v -s tests/basic_correctness/test_mem.py -k compressed_tensors_kv_scale
cd tests && VLLM_WORKER_MULTIPROC_METHOD=spawn pytest -v -s basic_correctness/test_mem.py
```

New case `test_deep_sleep_fp8_kvcache_compressed_tensors_kv_scale`, above the two
`…_mrv1` cases (the second builds its engine in the pytest parent and leaves its 0.9
KV pool mapped there, which a fork inherits). It asserts registration form, non-zero
values, exact equality with the pre-sleep copies, and greedy output/logprob parity
against a never-slept control, plus a `sleep(level=1)` negative control and an
anti-vacuity probe that the weights really were discarded and reloaded. Red arm =
revert only the fix. Exact equality is deliberate: an injected 4.657e-09
perturbation of these scales flipped a greedy decision at token 10.

Checkpoint `nm-testing/TinyLlama-1.1B-Chat-v1.0-kvcache-fp8-tensor` (static
`kv_cache_scheme`, so the scales are genuinely calibrated; already used by
`tests/quantization/test_compressed_tensors.py`). Measured floor 2.46 + 0.69 GiB, so
`gpu_memory_utilization=0.6` fits the `h200_18gb` slice this file runs on — no
`large_gpu_mark`, **not skipped in CI**.

## Test result (H20)

| arm | rc | result |
|---|---|---|
| pristine tree | **1** | `1 failed, 13 deselected, 17 warnings in 16.37s` — assertion names all 66 lost scales |
| with this fix | **0** | `1 passed, 13 deselected, 17 warnings in 23.18s` |
| with this fix, 10.8 GiB budget (≡ 18 GiB slice @ util 0.6) | **0** | `1 passed, 13 deselected, 17 warnings in 22.16s` |
| whole file as CI invokes it | **0** | `14 passed, 41 warnings in 221.09s` |

With the fix, output is **bit-exact against the never-slept control** (36/36 rounds,
324-tensor census 0 zeroed / 0 pointer moves), the noise floor is 0, and
`sleep(level=1)` is a clean negative control. Lint on the two changed files:
`ruff check` / `ruff format --diff` / `py_compile` clean and
`pre-commit run --files <both>` 13 Passed / 0 Failed — including
`check-forbidden-imports`, which is why the test resolves the hub id with the shared
`hf_api()` helper.

<details>
<summary>Full characterization on the 8B checkpoint (96/96 scales, NaN logprobs, FLASHINFER, CUDA graphs, TP=2, ATTN_HEAD)</summary>

8B `nm-testing/Llama-3.1-8B-Instruct-KV-FP8-tensor-static_minmax`, greedy with
logprobs: unfixed, **96/96** parameter-form scales read back `0.0` while **0/32**
buffer-form did and all 34,560 logprobs are NaN; with the fix, **36/36 rounds
bit-exact** vs the reload-only control, 324-tensor census 0 zeroed / 0 pointer
moves. Noise floor 0 (no-sleep reload 36/36 bit-exact; 22 pre-sleep references
across 22 processes identical); `sleep(level=1)` clean (0/128 changed). FLASHINFER
reproduces it identically, so it is not backend-specific; with CUDA graphs on, wake
re-captures nothing; at TP=2 exactly those 96 names flip form; ATTN_HEAD (`[8,1,1]`)
covered; old-format `quant_method="fp8"` unaffected at these scales.

</details>

## Model evaluation

GSM8K, full 1319-question test set, vLLM's own harness
(`tests/evals/gsm8k/gsm8k_eval.py` — `_build_gsm8k_prompts` / `get_answer_value` /
`_score_gsm8k`; 5-shot, greedy, `max_tokens=256`, seed 0) on
`nm-testing/TinyLlama-1.1B-Chat-v1.0-kvcache-fp8-tensor` with
`kv_cache_dtype=fp8`:

| arm | accuracy | invalid | vs baseline |
|---|---|---|---|
| A — never slept | 2.65% (35/1319) | 0.6% | — |
| B — level=2 → wake(weights) → weights-only reload → wake(kv_cache), **with fix** | 2.65% (35/1319) | 0.6% | **1319/1319 byte-identical output** |
| C — same sequence, **without fix** | **0.00% (0/1319)** | **100%** | 0/1319 |

Absolute accuracy is low because TinyLlama-1.1B-Chat is weak at 5-shot GSM8K; the
signal is that B reproduces A byte-for-byte while C collapses (all 66
`_q/_k/_v_scale` read back `0.0`, every response runs to `max_tokens`, most decoding
to an empty string). Both sleeping arms assert the weights pool really was discarded
and the reload really restored 135 parameters bit-exactly, so the delta can only
come from the scales. A 500-question subset gave the same picture (500/500).

## Scope

The documented level-2 recovery -- `collective_rpc("reload_weights")` -- self-heals, and
this PR does not claim otherwise: the layerwise path re-creates the `k_scale`/`v_scale`
placeholders before replaying the checkpoint, so `process_weights_after_loading()` runs
again (measured: green with and without the fix). What does not self-heal is anything
that skips it: `wake_up(tags=["weights"])` with no reload at all,
`reload_weights(is_checkpoint_format=False)` (the kernel-format branch, `param.copy_()`
only, next to the `# TODO: buffers?`), and a third-party `model.load_weights()`. There
the `named_buffers()` snapshot is the only net, which is the level-2 contract in
`docs/features/sleep_mode.md` ("the model's buffers are kept in CPU") and the invariant
`test_static_model_tensors_survive_level2_restore` asserts at the same point. Measured
matrix and the exact reason a plain reload cannot restore them: follow-up comment.

## Platforms

Verified on CUDA/H20 only. The AMD mirror of `basic_correctness` collects this test
(`requires_fp8` is true on CDNA, sleep mode is available on MI3xx) and two assertions are
platform-sensitive: `discarded["all_zero"]`, which relies on the driver handing back zeroed
pages, and the bit-exact logprob comparison. If the mirror goes red we will relax those two
(tolerance / platform guard) and keep the buffer-form assertion, which is the point of the test.

## Not covered (disclosed)

- W8A8 `weight_scale` / `input_scale` are `nn.Parameter` too and are lost by the
  same mechanism: on `neuralmagic/Meta-Llama-3-8B-Instruct-FP8-KV`, **449 of 580**
  censused tensors read back `0.0` after level=2 + a weights-only reload (256
  weight/input scales, 129 weights, 64 norms), while its 128 KV/attention scales are
  buffers and survive. Needs the worker-level net above — separate change.
- The test pins `VLLM_USE_V2_MODEL_RUNNER=0`, like the two adjacent `…_mrv1` cases.
  Both runners share `Worker.sleep()` and `_sleep_saved_buffers`, but we have not
  measured the V2 runner and do not claim it here.
- TP=2 shows a pre-existing per-engine output bimodality (2 of 24 snapshots; same
  signature with `--disable-custom-all-reduce`, also in the unfixed control arm) —
  unrelated to sleep or to this patch.

### CODE REVIEW

**Comment by AlanFokCo** [comment]
Measured the three recovery paths after `sleep(level=2)` on an H20 (CUDA, driver 580.126.09), with `nm-testing/TinyLlama-1.1B-Chat-v1.0-kvcache-fp8-tensor` (22 layers -> 88 attention scale names).

These numbers come from a throwaway probe file that is **not** part of this PR's diff -- the test in this PR is unchanged. Two trees, one probe file byte-identical on both; `diff -rq` between the trees reports only three differences (the quant-method file, the shipped `test_mem.py`, and a `.ruff_cache`), and the trainer-form payload's sorted name list hashes to the same value on both (`fa512f57…`), so the quant-method file is the only variable that can matter here.

| after `sleep(level=2)` + `wake_up(tags=["weights"])` | without this fix | with this fix |
|---|---|---|
| no reload at all | **66/88 scales read back 0.0, every logprob NaN, first tokens `[0, 0, 0, 0]` instead of `[13, 29902, 626, 2691]`** | 0 zeroed, values intact |
| `reload_weights(is_checkpoint_format=False)` with a trainer-form payload (`named_parameters()` minus the internal scale names -- a trainer does not own them) | **same 66 zeroed / NaN**, while a >1M-element weight probe comes back bit-exactly (same float64 checksum before and after), so the reload itself is not the variable | 0 zeroed, output bit-identical |
| `collective_rpc("reload_weights")` -- the recovery documented in `docs/features/sleep_mode.md` | 0 zeroed, bit-identical: **self-heals** | 0 zeroed, bit-identical |

Row 1 needs no assumption about payload shape at all, which is why it is the one I would point at: with nothing reloaded, the scales are simply gone. 66 = 22 layers x `_q/_k/_v_scale`; `_prob_scale`, which is never rebound, survives on both trees. `data_ptr()` of all 88 scales is unchanged across the sleep/wake in all six runs on both trees, so the fix relocates nothing.

Why a plain reload cannot put them back: `get_cache_scale_mapper()` maps the checkpoint's `self_attn.{k,q,v}_scale` onto `self_attn.attn.{k,q,v}_scale` (`vllm/model_executor/layers/quantization/base_config.py`, default `(?<!\.attn)\.([qkv])_scale$` rule) -- the placeholder that `create_weights()` made in this file -- and `process_weights_after_loading()` then `del`s it. Those 66 keys become unexpected on any later load, and `_ignore_unexpected_suffixes` in `base_config.py` lists `.q_scale`/`.k_scale`/`.v_scale`, so `AutoWeightsLoader` skips them silently (debug log only) -- there is not even a warning. The layerwise path is the exception because `_reload_attention_scales()` re-creates the placeholders before replaying the weights (`vllm/model_executor/model_loader/reload/layerwise.py`), which is the self-heal in row 3. For reference, this checkpoint carries exactly those 66 scale tensors (333 keys total) and its `kv_cache_scheme` in `config.json` holds no numeric scales, so the calibrated values can only come from them.

Also worth noting why the existing level-2 invariant test does not catch this today, for two independent reasons: `_get_sleep_tensor_targets()` in `tests/model_executor/test_sleep_mode_tensor_ownership.py` has no KV scales in its target list, and `_save_dummy_weights()` there snapshots unfiltered `named_parameters()`, which currently *includes* `_k_scale` precisely because it is a Parameter -- so `_reload_dummy_weights()` copies it back and masks the loss.

**Comment by AlanFokCo** [comment]
Update in bde294c5b189193a942eeb49c73992fb3cfb46a6.

Updated the regression test after rebasing onto the current PR head. The quantization fix itself is unchanged in this update.

The test now verifies that level-2 sleep unmaps every weights allocation without a CPU backup, corroborated by the increase in device free memory. It no longer reads discarded weights or assumes newly mapped pages contain zeros. It still checks restored scale values, a weight checksum, and generated output after the weights-only reload. Callable serialization is enabled only for the trusted local test probes; the CuMem-specific test explicitly skips XPU.

Validation on an H20, driver 580.126.09, PyTorch 2.13.0+cu130, using `nm-testing/TinyLlama-1.1B-Chat-v1.0-kvcache-fp8-tensor`:

```bash
python -m pytest tests/basic_correctness/test_mem.py -q -s \
  -k compressed_tensors_kv_scale
```

- Updated PR tree: **1 passed, 13 deselected**.
- Same test with only the quantization fix reverted: **1 failed as expected**, detecting the scale registration regression.
- Pre-commit checks passed for the changed file.

One correction to the terminology in my earlier comment: matching a float64 checksum is a checksum check, not proof of byte-for-byte equality. Likewise, zero-valued discarded scales were an observation in that experiment, not a guarantee about newly mapped memory.

This update and its local validation used AI assistance; these results are separate from upstream CI.
